dali/python/nvidia/dali/shared_batch.py

```diff
@@ -0,0 +1,221 @@
+                   writer.meta_data_size)
+
+
+def deserialize_sample(buffer: shared_mem.SharedMem, sample):
+    if isinstance(sample, SampleMeta):
+        offset = sample.offset
+        buffer = buffer.buf[offset:offset + sample.nbytes]
+        return np.ndarray(sample.shape, dtype=sample.dtype, buffer=buffer)
+    if isinstance(sample, (tuple, list,)):
+        return type(sample)(deserialize_sample(buffer, part) for part in sample)
+    return sample
+
+
+def deserialize_sample_meta(buffer: shared_mem.SharedMem, shared_batch_meta: SharedBatchMeta):
+    """Deserialize SampleMeta from memory based on SharedBatchMeta.
+    Deserialized SampleMeta can be used to reconstruct the batch data by passing it
+    to the `deserialize_batch`.
+    """
+    sbm = shared_batch_meta
+    if sbm.meta_size == 0:
+        return []
+    pickled_meta = buffer.buf[sbm.meta_offset:sbm.meta_offset + sbm.meta_size]
+    samples_meta = pickle.loads(pickled_meta)
+    return samples_meta
+
+
+def deserialize_batch(buffer: shared_mem.SharedMem, samples):
+    """Deserialize samples from the smem buffer and SampleMeta descriptions.
+
+    Parameters
+    ----------
+    buffer : shared_mem.SharedMem
+        Buffer with serialized sample data
+    samples : List of (idx, SampleMeta) or list of (idx, tuple of SampleMeta).
+        Metadata describing the samples
+
+    Returns
+    -------
+    List of (idx, numpy array) or (idx, tuple of numpy arrays)
+        List of indexed deserialized samples
+    """
+    return [(idx, deserialize_sample(buffer, sample)) for (idx, sample) in samples]
+
+
+class SharedMemChunk:
+    """Simple wrapper around shared memory chunks. Most importantly adds mem_chunk_id used
+    to identify chunks in the communication between parent and worker process
+    (file descriptors cannot serve this purpose easily as the same memory chunk
+    can have different fds in both processes).
+    """
+
+    def __init__(self, mem_chunk_id: str, capacity: int):
+        # mem_chunk_id must be unique among all workers and callbacks in the pool,
+        # used to identify shared memory chunks in the communication between processes
+        self.mem_chunk_id = mem_chunk_id
+        self.shm_chunk = shared_mem.SharedMem.allocate(capacity)
+        self.capacity = capacity
+
+    def resize(self, new_capacity=None):
+        capacity = self.capacity
+        new_capacity = new_capacity or 2 * capacity
+        self.shm_chunk.resize(new_capacity, trunc=True)
+        self.capacity = new_capacity
+
+    def close(self):
+        # Memory fd and mapping will be freed automatically when shm pybind wrapper gets
+        # garbage collected anyway, but you can be so nice to free them as soon as you know
+        # you won't use them
+        self.shm_chunk.close()
+
+
+def _to_numpy(sample):
+    if isinstance(sample, np.ndarray):
+        return sample
+    if types._is_mxnet_array(sample):
+        return sample.asnumpy()
```

Comment:
For paddle paddle it should just be enough `np.array(sample)`.

Reply:
But we don't have is pddle_tensor method....

Reply:
I'm going to write a list of gaps in ExternalSource we can close after this PR.